test/es-module/test-esm-dynamic-import.js

```diff
@@ -0,0 +1,96 @@
+// Flags: --experimental-modules --harmony-dynamic-import
+'use strict';
+const common = require('../common');
+const assert = require('assert');
+const { URL } = require('url');
+const vm = require('vm');
+
+common.crashOnUnhandledRejection();
+
+const relativePath = './test-esm-ok.mjs';
+const absolutePath = require.resolve('./test-esm-ok.mjs');
+const targetURL = new URL('file:///');
+targetURL.pathname = absolutePath;
+
+function expectErrorProperty(result, propertyKey, value) {
+  Promise.resolve(result)
+    .catch(common.mustCall(error => {
+      assert.equal(error[propertyKey], value);
+    }));
+}
+
+function expectMissingModuleError(result) {
+  expectErrorProperty(result, 'code', 'ERR_MISSING_MODULE');
+}
+
+function expectInvalidUrlError(result) {
+  expectErrorProperty(result, 'code', 'ERR_INVALID_URL');
+}
+
+function expectInvalidProtocolError(result) {
+  expectErrorProperty(result, 'code', 'ERR_INVALID_PROTOCOL');
+}
+
+function expectInvalidContextError(result) {
+  expectErrorProperty(result, 'message', 'invalid_context');
+}
+
+function expectOkNamespace(result) {
+  Promise.resolve(result)
+    .then(common.mustCall(ns => {
+      // Can't deepStrictEqual because ns isn't a normal object
+      assert.deepEqual(ns, { default: true });
+    }));
+}
+
+function expectFsNamespace(result) {
+  Promise.resolve(result)
+    .then(common.mustCall(ns => {
+      assert.equal(typeof ns.default.writeFile, 'function');
+    }));
+}
+
+expectOkNamespace(import(relativePath));
+
+expectFsNamespace(import("fs"));
+
+expectInvalidProtocolError(import("node:fs"));
+
+// Fails, no valid referrer (indirect eval) & relative path
+expectMissingModuleError(eval.call(null, `import("${relativePath}")`));
+
+// Fails, no valid referrer (direct eval) & relative path
+expectMissingModuleError(eval(`import("${relativePath}")`));
+
+// Succeeds, doesn't need a valid referrer
+expectFsNamespace(eval.call(null, 'import("fs")'));
+
+// Fails because referrer is undefined (and loader/search.js rejects it)
+// TODO(jkrems): This doesn't seem like expected behavior
+expectMissingModuleError(eval.call(null, `import("${targetURL}")`));
+
+// Succeeds
+expectFsNamespace(vm.runInThisContext('import("fs")'));
+
+// Succeeds because it's got an valid base url
+expectOkNamespace(vm.runInThisContext(`import("${relativePath}")`, {
+  filename: __filename,
+}));
+
+// Rejects because it's got an invalid referrer URL
+expectInvalidUrlError(vm.runInThisContext(`import("${relativePath}")`));
+
+// Rejects because it's got an invalid referrer URL
+expectInvalidUrlError(vm.runInThisContext(`import("${targetURL}")`));
```

Comment:
Is this rejecting an absolute file url?
```js
expectInvalidUrlError(vm.runInThisContext(`import("${targetURL}")`))
```

Reply:
@jdalton It seems like it.

Reply:
That seems really odd. Absolute paths don't have ambiguity like relative ones so it seems weird to error in this scenario (when they normally don't error).

Reply:
`vm` is somewhat of a special case because it doesn't really have an equivalent in browsers. But since I can't find a single case in my browsers testing where even relative imports don't work, I guess it would make sense to align `vm.runInThisContext` with `eval` / `new Function` / etc.?

Originally I thought that allowing `import("file:///absolute.mjs")` would poke a hole into `vm.runIn{This,New}Context` that doesn't exist today so I was hesitant to enable it. It's less of an issue in `vm.runInThisContext` where `process.mainModule.require` is accessible but it did seem like a bad precedent to expose it.

Reply:
[`vm` already lets you specify the source origin](https://github.com/nodejs/node/blob/d597317a209fcf9e0c233093b2adc8f71129b9ef/src/node_contextify.cc#L663). Is it too much to ask people to set that?

Reply:
>  Is it too much to ask people to set that?

I think `vm.runInThisContext('import(someAbsolutePath)')` should "just work". No extra options/config needed.

Reply:
I'm not convinced it should just work if you don't specify where the `import` is coming from. What do loader hooks see for example? I can't put `null` since that would be the same the module/script referrer for multiple module records.

Reply:
I need to think on if people need to include extra metadata for their loading like CORS via https://developer.mozilla.org/en-US/docs/Web/HTML/Element/script#attr-crossorigin which also affects anything in this vein of duplicate referrers.

Reply:
> I'm not convinced it should just work if you don't specify where the import is coming from. What do loader hooks see for example?

Have it get context similar to how `require` does. *Something to keep in mind when making these choices is every time we make ESM more cumbersome its another reason to stay on CJS.

Reply:
@jdalton `require` is not available by default in a `vm.Script`. You would need to pass one in.

Reply:
> `require` is not available by default in a `vm.Script`. You would need to pass one in.

I get it. I [showed an example](https://github.com/nodejs/node/pull/15713#discussion_r149586177) earlier of simply assigning `require` to a local var. The gist is the same though.

Reply:
@bmeck For `runInThisContext`, you *can* pretty reliably get a `require` if you know the absolute path of the thing you're trying to import (`process.mainModule.require("<absolute path")`). A local var won't help but anything global (like `process.mainModule` or @jdalton's example above) works.